File: sdk/x402_middleware.ts
/**
 * TRACE API — TypeScript x402 middleware
 *
 * Drop-in for any Express / Hono / Node.js x402 server.
 * Checks agent trust score before processing payment.
 */

const TRACE_API_URL = "https://api.trace.dev";

interface TraceScoreResult {
  provider_id: string;
  score: number;
  routing_decision: "ROUTE" | "ROUTE_WITH_CAUTION" | "HOLD" | "INVESTIGATE";
  components: {
    lcb: number;
    default_risk: number;
    cost_norm: number;
    trust_net: number;
    cap_match: number;
    sybil_risk: number;
    clique_penalty: number;
  };
  flags: string[];
  explanation: string;
  latency_ms: number;
  version: string;
}

interface TRACEMiddlewareOptions {
  apiKey: string;
  minScore?: number;
  apiUrl?: string;
}

export class TRACEMiddleware {
  private apiKey: string;
  private minScore: number;
  private apiUrl: string;

  constructor(options: TRACEMiddlewareOptions) {
    this.apiKey = options.apiKey;
    this.minScore = options.minScore ?? 0.35;
    this.apiUrl = options.apiUrl ?? TRACE_API_URL;
  }

  /**
   * Call before processing x402 payment.
   * Returns score result or throws an error with status 402 if agent is untrusted.
   */
  async check(
    agentWallet: string,
    jobCapability: string,
    priceUsdc: number
  ): Promise<TraceScoreResult> {
    const resp = await fetch(`${this.apiUrl}/v1/score`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-API-Key": this.apiKey,
      },
      body: JSON.stringify({
        provider_id: agentWallet,
        job: {
          capability: jobCapability,
          price_usdc: priceUsdc,
        },
        provider_history: {},
        graph_context: {},
      }),
      signal: AbortSignal.timeout(5000),
    });

    const result: TraceScoreResult = await resp.json();

    if (
      result.routing_decision === "HOLD" ||
      result.routing_decision === "INVESTIGATE"
    ) {
      const error = new Error(
        `Agent trust score ${result.score.toFixed(2)} below threshold ${this.minScore}`
      );
      (error as any).status = 402;
      (error as any).details = {
        error: "agent_untrusted",
        score: result.score,
        flags: result.flags,
      };
      throw error;
    }

    return result;
  }
}

/**
 * Express middleware factory.
 *
 * Usage:
 *   import { traceGate } from "./x402_middleware";
 *   app.use("/agent", traceGate({ apiKey: "sk_trace_..." }));
 */
export function traceGate(options: TRACEMiddlewareOptions) {
  const trace = new TRACEMiddleware(options);

  return async (req: any, res: any, next: any) => {
    const agentWallet =
      req.headers["x-payment-sender"] || req.headers["x-agent-wallet"];
    const capability = req.body?.capability ?? "default";
    const price = req.body?.price_usdc ?? 0.01;

    if (!agentWallet) {
      return next(); // No wallet header — skip trust check
    }

    try {
      const result = await trace.check(agentWallet, capability, price);
      req.traceScore = result;
      next();
    } catch (err: any) {
      res.status(err.status || 402).json({
        error: "agent_untrusted",
        score: err.details?.score,
        flags: err.details?.flags,
        message: err.message,
      });
    }
  };
}

File: sdk/x402_middleware.py
"""
TRACE API — Python x402 middleware

Drop-in for any FastAPI or Flask x402 server.
Checks agent trust score before processing payment.
"""

import httpx
from fastapi import Request, HTTPException

TRACE_API_URL = "https://api.trace.dev"  # or self-hosted


class TRACEMiddleware:
    """
    x402-compatible trust gate for AI agent marketplaces.

    Usage:
        trace = TRACEMiddleware(api_key="sk_trace_...")

        @app.post("/your-agent-endpoint")
        async def handle_job(request: Request):
            agent_wallet = request.headers.get("X-Payment-Sender")
            score = await trace.check(agent_wallet, "summarize", 0.05)
            # proceed with job ...
    """

    def __init__(
        self,
        api_key: str,
        min_score: float = 0.35,
        api_url: str = TRACE_API_URL,
    ):
        self.api_key = api_key
        self.min_score = min_score
        self.api_url = api_url

    async def check(
        self,
        agent_wallet: str,
        job_capability: str,
        price_usdc: float,
    ) -> dict:
        """
        Call before processing x402 payment.
        Returns score dict or raises HTTPException(402) if agent is untrusted.
        """
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{self.api_url}/v1/score",
                json={
                    "provider_id": agent_wallet,
                    "job": {
                        "capability": job_capability,
                        "price_usdc": price_usdc,
                    },
                },
                headers={"X-API-Key": self.api_key},
                timeout=5.0,
            )

        result = resp.json()

        if result["routing_decision"] in ("HOLD", "INVESTIGATE"):
            raise HTTPException(
                status_code=402,
                detail={
                    "error": "agent_untrusted",
                    "score": result["score"],
                    "flags": result["flags"],
                    "message": (
                        f"Agent trust score {result['score']:.2f} "
                        f"below threshold {self.min_score}"
                    ),
                },
            )

        return result

    async def report_event(
        self,
        agent_wallet: str,
        success: bool,
        buyer_wallet: str = None,
        job_capability: str = None,
        price_usdc: float = None,
    ):
        """
        Report the outcome of a job back to TRACE to update the agent's trust score.
        """
        async with httpx.AsyncClient() as client:
            await client.post(
                f"{self.api_url}/v1/events",
                json={
                    "provider_id": agent_wallet,
                    "buyer_id": buyer_wallet,
                    "success": success,
                    "capability": job_capability,
                    "price_usdc": price_usdc,
                },
                headers={"X-API-Key": self.api_key},
                timeout=5.0,
            )
